# Introduction to Numpy

A library for arrays and scientific computations in Python. [Also Check Sympy Here.](https://www.sympy.org/en/index.html)

### Disclaimer

1. We don't need to import the modules everytime. This is just for demonstration purposes.
2. This notebook contains just enough features of numpy and linear algebra to get started with machine learning fundamentals.
3. cos-1 means cos inverse, sin-1 means sin inverse, and so on.
4. x^2 means x squared.
5. If a vector is denoted like [[a], [b]] and not [a, b] it means it is 2 x 1 vector (a column) and not a row.
6. "==" should be read as "must be equal to" in notes.
7. These notes do not serve as an alternative to a full course in university, school, or college. These notes only work as refresher guide for those who want to learn machine learning.
8. Expand each section to see more.

Special Thanks to: [YouTube Channel 3Blue1Brown](https://www.youtube.com/watch?v=fNk_zzaMoSs&list=PLZHQObOWTQDPD3MizzM2xVFitgF8hE_ab)


## Importing the modules and some basic info


In [13]:
# Importing the module as np (generally accepted abbreviation)
import numpy as np

np.set_printoptions(suppress=True)  # sets the print options to real numbers instead of using 'e' scientific notations.

In [14]:
# Print Numpy Version
import numpy as np

print("Numpy Version:", np.__version__)

Numpy Version: 1.26.2


In [15]:
# Print Numpy configuration info
import numpy as np

print("====================== Numpy Config ======================")
print(np.show_config())

====================== Numpy Config ======================
Build Dependencies:
  blas:
    detection method: pkgconfig
    found: true
    include directory: /opt/arm64-builds/include
    lib directory: /opt/arm64-builds/lib
    name: openblas64
    openblas configuration: USE_64BITINT=1 DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS=
      NO_LAPACK= NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP= SANDYBRIDGE MAX_THREADS=3
    pc file directory: /usr/local/lib/pkgconfig
    version: 0.3.23.dev
  lapack:
    detection method: internal
    found: true
    include directory: unknown
    lib directory: unknown
    name: dep4413533856
    openblas configuration: unknown
    pc file directory: unknown
    version: 1.26.2
Compilers:
  c:
    commands: cc
    linker: ld64
    name: clang
    version: 14.0.0
  c++:
    commands: c++
    linker: ld64
    name: clang
    version: 14.0.0
  cython:
    commands: cython
    linker: cython
    name: cython
    version: 3.0.5
Machine Information:
  build:
    cpu: aar

In [16]:
# Getting Help and Info
import numpy as np

np.info(np.add) # ---> Get information about add function of numpy.

add(x1, x2, /, out=None, *, where=True, casting='same_kind', order='K', dtype=None, subok=True[, signature, extobj])

Add arguments element-wise.

Parameters
----------
x1, x2 : array_like
    The arrays to be added.
    If ``x1.shape != x2.shape``, they must be broadcastable to a common
    shape (which becomes the shape of the output).
out : ndarray, None, or tuple of ndarray and None, optional
    A location into which the result is stored. If provided, it must have
    a shape that the inputs broadcast to. If not provided or None,
    a freshly-allocated array is returned. A tuple (possible only as a
    keyword argument) must have length equal to the number of outputs.
where : array_like, optional
    This condition is broadcast over the input. At locations where the
    condition is True, the `out` array will be set to the ufunc result.
    Elsewhere, the `out` array will retain its original value.
    Note that if an uninitialized `out` array is created via the default
    ``out

## Creating Arrays

There are three ways of creating arrays from numpy.

1. Use Python sequential objects to create numpy arrays.
2. Use Numpy's intrinsic functions to create arrays.
3. Copy an existing array.


### Creating Arrays from Python Objects


In [21]:
# Create array with python lists
import numpy as np

arr_list = np.array([1, 5, 7, 9, 4])
print(f'List Array: {arr_list}')

# Print length of an array
print("Length of list array: ", len(arr_list))

List Array: [1 5 7 9 4]
Length of list array:  5


In [22]:
# Create array with python tuples
import numpy as np

arr_tupl = np.array((1, 2, 8, 6))
print(f'Tuple Array: {arr_tupl}')

Tuple Array: [1 2 8 6]


In [23]:
# Create 2d array with python list
import numpy as np

arr_2d  = np.array([[1, 2], [3, 4]])
print("2D array:")
print(arr_2d)

# Check array dimensions
print(f'2D array dimension: {arr_2d.ndim}')

print("Length of 2D array: ", len(arr_2d))

2D array:
[[1 2]
 [3 4]]
2D array dimension: 2
Length of 2D array:  2


In [25]:
# Creating ndim array
import numpy as np

arr_ndim = np.array([1, 2, 3, 4, 5], ndmin=5)   # create a 5th dimensional array
print(f'5D Array: {arr_ndim} and Dimension: {arr_ndim.ndim}')

5D Array: [[[[[1 2 3 4 5]]]]] and Dimension: 5


#### Create Array and Define Type for Memory Optimization

Specifying data type explicitly allows numpy to save memory and make calculations more faster.

[Check all datatypes here](https://numpy.org/doc/stable/user/basics.types.html)


In [27]:
# Creating array and defining the type for optimizations
import numpy as np

ls = np.array([0, 11, 22, 33, 44, 55], dtype=np.uint16)
print(ls)

[ 0 11 22 33 44 55]


### Numpy Methods to Create Arrays


#### Arange method np.arange()

Syntax: np.arange(start, end, step, dtype=np.int8)</br>
</br>
The end is not inclued, mathematically something like this - [start, end)


In [28]:
# Create different types of numpy array with np.arange()
import numpy as np

arr = np.arange(1, 10) # start, end (not included)
print(f'Arange array: {arr}')

arr = np.arange(1, 10, dtype=np.int8)
print(f'Arange array {arr} with datatype {arr.dtype}')

arr = np.arange(1, 10, dtype=float)
print(f'Arange array {arr} with datatype {arr.dtype}')

arr = np.arange(1, 2, 0.1)
print(f'Arange array starting at 1, stopping at 1.9 with step 0.1 = {arr}')

Arange array: [1 2 3 4 5 6 7 8 9]
Arange array [1 2 3 4 5 6 7 8 9] with datatype int8
Arange array [1. 2. 3. 4. 5. 6. 7. 8. 9.] with datatype float64
Arange array starting at 1, stopping at 1.9 with step 0.1 = [1.  1.1 1.2 1.3 1.4 1.5 1.6 1.7 1.8 1.9]


#### Identity Matrix

Syntax: np.eye(rows, cols)

If the rows != cols then other elements are set to 0


In [106]:
# Create Identity Matrix with numpy
import numpy as np

idmat = np.eye(3)
print("3 x 3 Identity Matrix")
print(idmat)
print("\n")

print("3 x 3 Identity Matrix")
idmat = np.eye(3, 3)
print(idmat)
print("\n")

print("3 x 6 Identity Matrix with extra elements set to 0")
idmat = np.eye(3, 6)
print(idmat)

3 x 3 Identity Matrix
[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


3 x 3 Identity Matrix
[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


3 x 6 Identity Matrix with extra elements set to 0
[[1. 0. 0. 0. 0. 0.]
 [0. 1. 0. 0. 0. 0.]
 [0. 0. 1. 0. 0. 0.]]


#### Creating nD arrays

No direct method available. Use the reshape() function to reshape a 1D Array.

Syntax: arr.reshape(rows, cols, z, ...) ---> n arguments will shape the array in n dimensions.</br>
</br>
<strong>Note: </strong>rows _ cols _ z \* ... all dimensions product must be euqal to number of items in the array.


In [33]:
# Reshaping array
import numpy as np

mat = np.arange(1, 10).reshape(3, 3) # provide more int args to the reshape function to create more dimensions
print("A 2D 3x3 matrix:")
print(mat)
print("\n")

mat = np.arange(0, 16).reshape(2, 2, 2, 2)  # creates a 4d array
print("A 4D 2x2x2x2 matrix:")
print(mat)
print("\n")

A 2D 3x3 matrix:
[[1 2 3]
 [4 5 6]
 [7 8 9]]


A 4D 2x2x2x2 matrix:
[[[[ 0  1]
   [ 2  3]]

  [[ 4  5]
   [ 6  7]]]


 [[[ 8  9]
   [10 11]]

  [[12 13]
   [14 15]]]]




#### Diagonal Matrix

Syntax: np.diag([a, b, c])

Creates a diagonal matrix


In [37]:
# Creating a diagonal matrix
import numpy as np

diag = np.diag([1, 2, 3])
print("Simple diagonal matrix:")
print(diag)
print("\n")

diag = np.diag([1, 2, 3], 2)
print("A (3 + 2) * (3 + 2) diagonal matrix with all other values as 0:")
print(diag)
print("\n")

arr = np.array([[1, 2], [3, 4]])
print("Original matrix:")
print(arr)
diag = np.diag(arr)
print("Extracting the diagonal elements from a matrix:")
print(diag)
print("\n")

Simple diagonal matrix:
[[1 0 0]
 [0 2 0]
 [0 0 3]]


A (3 + 2) * (3 + 2) diagonal matrix with all other values as 0:
[[0 0 1 0 0]
 [0 0 0 2 0]
 [0 0 0 0 3]
 [0 0 0 0 0]
 [0 0 0 0 0]]


Original matrix:
[[1 2]
 [3 4]]
Extracting the diagonal elements from a matrix:
[1 4]




#### Filling Arrays

Creates an array and fills it with 0, 1, or random numbers.


In [38]:
# Zero filled arrays
import numpy as np

zeroes = np.zeros((3))  # Creates an 1D array
print(zeroes)

[0. 0. 0.]


In [39]:
# One filled arrays
import numpy as np

ones = np.ones((3, 3)) # Creates a 2D matrix
print(ones)

[[1. 1. 1.]
 [1. 1. 1.]
 [1. 1. 1.]]


In [42]:
# Random values filled arrays
from numpy.random import default_rng

mat = default_rng(3).random((2,3)) # Creates a 2x3 2D matrix
print(mat)

[[0.08564917 0.23681051 0.80127447]
 [0.58216204 0.09412864 0.43312694]]


### Copying Arrays

There are two types of copying of arrays in Numpy

1. Shallow copy (also called views) - these lets you see a view of the original array. Uses slice to create views.
2. Deep copy, creates a new array. Uses copy() to create a copy


In [45]:
# Shallow Copy or Views Demonstration
import numpy as np

a = np.array([1, 2, 3, 4, 5, 6])
print("Original array: ", a, "\n")

b = a[2:]
print("View of a: ", b)

print("Adding one two each element of b using array broadcasting\n")
b += 1  # adds one to each element in the array

print("View of a: ", b)
print("Original array a: ", a, "\n")

print("Chaging b[0], i.e., a[2] to 55")
b[0] = 55  # adds one to each element in the array

print("View of a: ", b)
print("Original array a: ", a)

Original array:  [1 2 3 4 5 6] 

View of a:  [3 4 5 6]
Adding one two each element of b using array broadcasting

View of a:  [4 5 6 7]
Original array a:  [1 2 4 5 6 7] 

Chaging b[0], i.e., a[2] to 55
View of a:  [55  5  6  7]
Original array a:  [ 1  2 55  5  6  7]


In [50]:
# Deep Copy or Views Demonstration
import numpy as np

a = np.array([1, 2, 3, 4, 5, 6])
print("Original array: ", a)

b = a[2:].copy() # deep copy
print("b: ", b, "\n")

print("Adding one two each element of b using array broadcasting\n")
b += 1  # adds one to each element in the array

print("b: ", b)
print("Original array a is still: ", a)

Original array:  [1 2 3 4 5 6]
b:  [3 4 5 6] 

Adding one two each element of b using array broadcasting

b:  [4 5 6 7]
Original array a is still:  [1 2 3 4 5 6]


## Array Indexing, Slicing and Printing

Array can be sliced and assigned to other variables just like lists. However the syntax is a bit different.</br>
</br>
Syntax: arr[start: stop: step, start: stop: step, ..., n] ---> for n dimension array


In [52]:
# Slicing a 1D array
import numpy as np

arr = np.arange(0, 10, 1, dtype=np.uint8)
print("Array: ", arr)
print("Slice [0: 7: 2]: ", arr[0: 7: 2])


Array:  [0 1 2 3 4 5 6 7 8 9]
Slice [0: 7: 2]:  [0 2 4 6]


In [54]:
# Slicing a 2D matrix
import numpy as np

mat = np.arange(1, 17, 1, dtype=np.uint8).reshape(4, 4)
print("Matrix: ")
print(mat)
print("\n")

print("Slice [0: 4: 2, 0: 2]: ")
print(mat[0: 4: 2, 0: 2])

Matrix: 
[[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]
 [13 14 15 16]]


Slice [0: 4: 2, 0: 2]: 
[[ 1  2]
 [ 9 10]]


## Array Operations

[Check all array manipulation methods here](https://numpy.org/doc/stable/reference/routines.array-manipulation.html)


### Stacking Arrays

Two types of stacking

1. Horizontal Stack or HStack - use the hstack() function
2. Vertical Stack or VStack - use the vstack() function


In [58]:
# Stacking arrays Horizontally
import numpy as np

a = np.zeros((3, 3))
b = np.ones((3, 3))

hs = np.hstack((a, b))

print("Horizontal stack")
print(hs)

Horizontal stack
[[0. 0. 0. 1. 1. 1.]
 [0. 0. 0. 1. 1. 1.]
 [0. 0. 0. 1. 1. 1.]]


In [59]:
# Stacking arrays Vertically
import numpy as np

a = np.zeros((3, 3))
b = np.ones((3, 3))

print(a)
print(b)

vs = np.vstack((a, b))

print("Vertical stack: ")
print(vs)

[[0. 0. 0.]
 [0. 0. 0.]
 [0. 0. 0.]]
[[1. 1. 1.]
 [1. 1. 1.]
 [1. 1. 1.]]
Vertical stack: 
[[0. 0. 0.]
 [0. 0. 0.]
 [0. 0. 0.]
 [1. 1. 1.]
 [1. 1. 1.]
 [1. 1. 1.]]


### Splitting Arrays

Two types of splitting

1. Horizontal Split or HSplit - use the hsplit() function
2. Vertical Split or VSplit - use the vsplit() function


In [61]:
# Vertical Split
import numpy as np

mat = np.arange(16).reshape(4, 4)
print("Original matrix:")
print(mat, "\n")

x, y = np.vsplit(mat, 2)    # the split number should be a divisor of row numbers

print("Upper half:")
print(x)

print("Lower half")
print(y)

Original matrix:
[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]
 [12 13 14 15]] 

Upper half:
[[0 1 2 3]
 [4 5 6 7]]
Lower half
[[ 8  9 10 11]
 [12 13 14 15]]


In [63]:
# Horizontal Split
import numpy as np

mat = np.arange(12).reshape(3, 4)
print("Original matrix:")
print(mat, "\n")

x, y = np.hsplit(mat, 2)    # the split number should be a divisor of col numbers

print("Left half:")
print(x)

print("Right half")
print(y)

Original matrix:
[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]] 

Left half:
[[0 1]
 [4 5]
 [8 9]]
Right half
[[ 2  3]
 [ 6  7]
 [10 11]]


### Concatenate

By default the axis = 0 means concatenate rows or vstack. </br>
The axis = 1 means concatenate cols or hstack.</br>
</br>
Length should be same on the chosen axis for a successful concatenation.


In [68]:
# Concatenating arrays or matrices

a = np.arange(4).reshape(2, 2)
b = np.arange(4, 8).reshape(2, 2)

print("a: ")
print(a, "\n")

print("b: ")
print(b, "\n")

print("Vertiacally Concatenated:")
print(np.concatenate((a, b), axis=0), "\n")

print("Horizontally Concatenated:")
print(np.concatenate((a, b), axis=1))

a: 
[[0 1]
 [2 3]] 

b: 
[[4 5]
 [6 7]] 

Vertiacally Concatenated:
[[0 1]
 [2 3]
 [4 5]
 [6 7]] 

Horizontally Concatenated:
[[0 1 4 5]
 [2 3 6 7]]


## Array Broadcasting

Automatically expand or stretch the smaller arrays to perform arithmetic operations.


In [78]:
# Broadcast addition (or subtraction)
import numpy as np

mat = np.arange(16).reshape(4, 4)
print("Original matrix:")
print(mat, "\n")

print("Adding two to each element in the matrix: ")
print(mat + 2, "\n")

print("Adding [1, 2, 3, 4] row array to every row of the original matrix:")
print(mat + np.array([1, 2, 3, 4]), "\n")

print("Adding [1, 2, 3, 4] col array to every col of the original matrix")
print(mat + np.array([1, 2, 3, 4]).reshape(4, 1))

Original matrix:
[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]
 [12 13 14 15]] 

Adding two to each element in the matrix: 
[[ 2  3  4  5]
 [ 6  7  8  9]
 [10 11 12 13]
 [14 15 16 17]] 

Adding [1, 2, 3, 4] row array to every row of the original matrix:
[[ 1  3  5  7]
 [ 5  7  9 11]
 [ 9 11 13 15]
 [13 15 17 19]] 

Adding [1, 2, 3, 4] col array to every col of the original matrix
[[ 1  2  3  4]
 [ 6  7  8  9]
 [11 12 13 14]
 [16 17 18 19]]


In [80]:
# Broadcast multiplication
import numpy as np

mat = np.arange(16).reshape(4, 4)
print("Original matrix:")
print(mat, "\n")

print("Multiplying two to each element in the matrix: ")
print(mat * 2, "\n")

Original matrix:
[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]
 [12 13 14 15]] 

Multiplying two to each element in the matrix: 
[[ 0  2  4  6]
 [ 8 10 12 14]
 [16 18 20 22]
 [24 26 28 30]] 



# Vectors with Numpy

## Introduction to Vectors

### Computer Science

In computer science, vectors are ordered lists of numbers to represent different values of the same entity (usually). Example, a house can be represented by values of it's different metrics such as no. of rooms, total area, no. of floors, price, etc.

It can be represented vertically or horizontally.</br>
[3, 1400, 2, 150000] where 3 is the no. of rooms, 1400 is the area in sq. ft, 2 is the number of floors and 150000 is the cost.</br>
The order of this sequence matters in calculations.
</br>

<strong>Note: </strong>Vectors in computer science always start from the origin (0, 0), unless otherwise stated.

### Mathematics

In Mathematics, vectors are lines on 2D, 3D, or nD space with a magnitude and a direction. <strong>In a 2D plane [x, y] represents a vector where tan-1(y / x) gives us the direction of the vector and sqrt(x^2 + y^2) gives the magnitude of the vector.</strong>


## Vector Arithmetic

Vectors can be added, subtracted, multiplied (dot and cross), etc just like real numbers.


### Addition of Vectors

Let's say a vector p(x, y) goes from point A to point B and a vector q(i, j) goes from point B to point C in a 2D space. The vector r(x+i, y+j) = p + q means going from point A to point C directly.</br>
</br>
In numpy a vector is an n x 1 entity. An easy way to create this vector out of python list is to use the reshape(n, 1).</br>
The vector will always be represented as an n x 1 entity except in some special cases.


In [79]:
# Addition of Vectors
import numpy as np

# Create two vectors
p = np.array([1, 2]).reshape(2, 1)
q = np.array([4, -7]).reshape(2, 1)

# Add those vectors
r = p + q

print("p:")
print(p)
print("q:")
print(q)
print("\n")
print("p + q:")
print(r)

p:
[[1]
 [2]]
q:
[[ 4]
 [-7]]


p + q:
[[ 5]
 [-5]]


### Distance between two Vectors

The distance r between a vector u(2, 1) and a vector v(8, 9) can be calculated in following ways:

1. L1 Method v - u = r(6, 8)
2. L2 Method r = sqrt((8-2)^2 + (9-1)^2) = 10
3. Angle between them: cos-1(u.v / (|u| . |v|)) where |u| and |v| denotes the magnitude of the vectors u and v respectively.

<strong>Note: </strong>For all notes, most of the time L2 methods will be used, unless stated otherwise.


In [80]:
# L1 Method
import numpy as np

# Create two vectors
u = np.array([2, 1]).reshape(2, 1)
v = np.array([8, 9]).reshape(2, 1)

# Find the distance between them using L1 method
dist = v - u

print("u:")
print(u)
print("v:")
print(v)
print("\n")
print("L1 Distance:")
print(dist)

u:
[[2]
 [1]]
v:
[[8]
 [9]]


L1 Distance:
[[6]
 [8]]


In [77]:
# L2 Method
import numpy as np

u = np.array([2, 1]).reshape(2, 1)
v = np.array([8, 9]).reshape(2, 1)

# Calculate the v - u and then normalise it to find the magnitude
dist = np.linalg.norm(v - u)   

print("u:")
print(u)
print("v:")
print(v)
print("\n")
print("L2 Distance:", dist)

u:
[[2]
 [1]]
v:
[[8]
 [9]]


L2 Distance: 10.0


In [81]:
# Angle between two vectors
import numpy as np

# Create two vectors - without reshaping to match the dimensions.
vector_u = np.array([3, 4])
vector_v = np.array([1, -1])

# Calculate the dot product
dot_product = np.dot(vector_u, vector_v)

# Calculate the magnitudes of the vectors
magnitude_u = np.linalg.norm(vector_u)
magnitude_v = np.linalg.norm(vector_v)

# Calculate the cosine of the angle
cosine_theta = dot_product / (magnitude_u * magnitude_v)

# Calculate the angle in radians
angle_radians = np.arccos(cosine_theta)

# Convert the angle to degrees
angle_degrees = np.degrees(angle_radians)

# Print the result
print("vector u:")
print(vector_u)
print("vector v:")
print(vector_v)
print("Angle between the vectors (in radians):", angle_radians)
print("Angle between the vectors (in degrees):", angle_degrees)

vector u:
[3 4]
vector v:
[ 1 -1]
Angle between the vectors (in radians): 1.7126933813990606
Angle between the vectors (in degrees): 98.13010235415598


### Product of two Vectors

There are two types of products of Vectors

1. Scaling: Scale a vector by a scalar quantity c.
2. Dot Product or Scalar Product: denoted by '.'
3. Cross Product or Vector Product: denoted by 'x'


#### Scaling a Vector

When a Vector v is multiplied by a scalar quantity c, the vector is scaled.


In [116]:
# Scaling a vector
import numpy as np

vector_v = np.array([1, 2, 3])
scaled_v = 3 * vector_v

print("Original vector:\t", vector_v)
print("Vector scaled by 3:\t", scaled_v)

Original vector:	 [1 2 3]
Vector scaled by 3:	 [3 6 9]


#### Dot Product or Scalar Product

Let's assume two vectors: u(1, 2, 3) and v(4, 5, 6)</br>
The dot product of two vectors u . v is (1 _ 2) + (2 _ 5) + (3 \* 6) = 30</br>
The Dot product of two vectors has following properties:

- The dot product of two vectors always gives a real number.
- If the two vectors are orthogonal, i.e., the angle between them is 90deg the dot product is always 0.
- It is because cos(90) = 0. Remember we found the angle between them using cos-1 and cos-1(90deg) = 0.
- The dot product is symmetrical i.e., u . v == v. u
- The dot product is distributive over vector addition i.e., r . (u + v) = r . u + r . v
- The dot product of a vector to itself i.e., u . u is |u|^2 (square of it's magnitude).


In [85]:
# Dot product of two vectors
import numpy as np

# Create two vectors - without reshaping to match the dimensions.
vector_u = np.array([3, 4])
vector_v = np.array([1, -1])

# Calculate the dot product
result = np.dot(vector_u, vector_v)

# Display the result
print("Vector u:", vector_u)
print("Vector v:", vector_v)
print("Dot product:", result)

Vector u: [3 4]
Vector v: [ 1 -1]
Dot product: -1


#### Cross Product or Vector Product

Let's assume two vectors u(a, b, c) and v(x, y, z)</br>
The cross product of two vectors u x v can be denoted by:</br>
</br>
| i&nbsp; j &nbsp;k |</br>
| a b c |</br>
| x y z |</br>
</br>
Which is: (bz – cy)i – (az – cx)j + (ay – bx)k = (bz – cy)i + (cx – az)j + (ay – bx)k</br>
The cross product of two vectors has following properties:

- The cross product is perpendicular to both the original vectors. And hence, u . (u x v) == v . (u x v) == 0
- The cross product of two parallel vectors is 0
- Non cumulative i.e., u x v != v x u but u x v == -(v x u)
- Distributive over vector addition i.e., r x (u + v) = r x u + r x v
- The angle relations is |u x v| = |u| . |v| . sin(theta)


In [86]:
import numpy as np

def calculate_cross_product(vector_u, vector_v):
    cross_product = np.cross(vector_u, vector_v)
    return cross_product

# Example vectors (3D vectors for the cross product)
vector_u = np.array([3, 4, 0])
vector_v = np.array([1, -1, 2])

# Calculate the cross product
result = calculate_cross_product(vector_u, vector_v)

# Display the result
print("Vector u:", vector_u)
print("Vector v:", vector_v)
print("Cross product:", result)

Vector u: [3 4 0]
Vector v: [ 1 -1  2]
Cross product: [ 8 -6 -7]


# Matrices with Numpy (and Sympy)

Matrices are rectangular arrays of numbers, symbols, or expressions arranged in rows and columns. They are fundamental mathematical objects that find applications in various fields, including linear algebra, physics, computer science, and engineering.

A matrix is typically denoted by a capital letter, and its entries are referred to as elements. <strong>The size or order</strong> of a matrix is specified by its number of rows and columns. For example, an "m x n" matrix has m rows and n columns.</br>
</br>
A matrix can have N dimensions and it is obvious when the order of the matrix is written. m x n x ... (N-1)th dimension length x Nth dimension length is an N dimensional matrix.</br>
</br>
We denote a matrix as a square or rectangle entity but here we will denote them like this: A = [[1, 2], [4, 5], [7, 8]] is a 3 x 3 matrix. It can be easily observed that outer list has 3 elements but inner list has two elements. <strong>The outer list represents rows, and the inner list represents cols. This analogy can be generalised to N dimensions</strong> </br>

- A matrix with only one row R = [[1, 2, 3]] called a row vector.
- A matrix with only one column C = [[1], [2], [3]] called col vector.
- A(i, j) denotes element of the matrix A at ith row and jth column.

<strong>Note 1: </strong>We will use sympy rarely for some computations.</br>
<strong>Note 1: </strong>We will use asmatrix() function of numpy to differentiate between matrix and an array. It is not necessary but it gives additional matrix functions to work with.


In [92]:
# Making a matrix
import numpy as np

matAsArray = np.array([[1, 2, 3, 0], [4, 5, 6, 1], [7, 8, 9, 2]])
mat = np.asmatrix(matAsArray)

print("Matrix A:")
print(mat)

print("Order of matrix A:", mat.shape)

Matrix A:
[[1 2 3 0]
 [4 5 6 1]
 [7 8 9 2]]
Order of matrix A: (3, 4)


## Special Types of Matrices

There can be different types of Matrices. Some of them are:

1. <strong>Square Matrix: </strong>If the order of a matrix A is m x n where m == n, the matrix is said to be square. Simply, no. of rows == no. cols in the matrix.
2. <strong>Identity Matrix: </strong>A matrix is said to be identity matrix I if for a matrix A, A x I = A. I(i, j) = 1 when i == j and 0 when i != j.
3. <strong>Diagonal Matrix: </strong>A matrix is said to be a diagonal matrix D if D(i, j) = x when i == j and 0 when i != j where x is any real number.
4. <strong>Triangular Matrix: </strong>A matrix whose only upper or lower triangle is filled with a number and rest is 0. It is usually the upper right half.

<strong>Note: </strong>There are other types of special matrices which we will cover later, while learning relevant concepts.


In [118]:
# Square Matrix
import numpy as np

matAsArray = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
mat = np.asmatrix(matAsArray)

print("Matrix A:")
print(mat)

print("Order of matrix A:", mat.shape) # see that m == n

Matrix A:
[[1 2 3]
 [4 5 6]
 [7 8 9]]
Order of matrix A: (3, 3)


In [119]:
# Identity Matrix
import numpy as np

i = np.asmatrix(np.eye(3, 3))

print("Identity matrix:")
print(i)

Identity matrix:
[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


In [120]:
# Diagonal Matrix
import numpy as np

diag = np.asmatrix(np.diag([1, 2, 3]))

print("Diagonal Matrix:")
print(diag)

Diagonal Matrix:
[[1 0 0]
 [0 2 0]
 [0 0 3]]


In [121]:
# Triangular Matrix
import numpy as np

trilist = [[3 if i <= j else 0 for j in range(3) ] for i in range(3)] # list comprehension for an upper right triangular matrix.
mat = np.asmatrix(np.array(trilist))

print("Upper Right Triangular Matrix:")
print(mat)


Upper Right Triangular Matrix:
[[3 3 3]
 [0 3 3]
 [0 0 3]]


## Matrix Arithmetic

Matrices can be added, subtracted, multiplied just like real numbers.


### Addition of Matrices

Matrices can be added, subtracted just like real numbers. However, there is a catch. The order must be same.


In [117]:
# Addition of two matrices
import numpy as np

# Create two matrices A and B
matA = np.asmatrix(np.array([
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9]
]))
matB = np.asmatrix(np.diag([1, 2, 3]))    # creates an diagonal matrix of order 3 x 3

print("Matrix A:")
print(matA)
print("\n")

print("Matrix B:")
print(matB)
print("\n")

print("Sum of the matrices:")
print(matA + matB)

Matrix A:
[[1 2 3]
 [4 5 6]
 [7 8 9]]


Matrix B:
[[1 0 0]
 [0 2 0]
 [0 0 3]]


Sum of the matrices:
[[ 2  2  3]
 [ 4  7  6]
 [ 7  8 12]]


### Product of the Matrices

Marices can be multiplied just like real numbers. However, there is a catch:
If a matrix A of order m1 x n1 has to be multiplied with a matrix B of order m2 x n2, n1 == m2 i.e., number of columns of A must be equal to number of rows of B. The order of the resultant matrix would be m1 x n2</br>
</br>
There are 3 types of product of matrices:

1. Scaling
2. Vector Multiplication
3. Matrix Multiplication: Matrix A can be multiplied with a matrix B.


#### Scalar Multiplication or Scalaing

When a real number c is multiplied to a matrix A, A gets scaled. Meaning: c _ A == c _ A(i, j) for all values of i and j.


In [138]:
# Scaling a matrix
import numpy as np

A = np.asmatrix(np.array([
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9]
]))
c = 5

print("Original matrix A:")
print(A)
print("\n")

print("Matrix after scaling by 5:")
print(c * A)

Original matrix A:
[[1 2 3]
 [4 5 6]
 [7 8 9]]


Matrix after scaling by 5:
[[ 5 10 15]
 [20 25 30]
 [35 40 45]]


#### Vector Multiplication

Matrix A can be multiplied with a vector v. However, the vector is treated like a matrix and the order constraint must be followed.
<strong>Note: </strong>A x v != v x A


In [137]:
# Multiplication with a row vector
import numpy as np

r = np.array([1, 2])   # row vector 1 x 2
A = np.asmatrix(np.array([ # matrix of 2 x 3
    [4, 5, 6],
    [7, 8, 9]
]))
res = np.matmul(r, A) # matrix multiplication

print("Original matrix A:")
print(A)
print("Order of A: ", A.shape)
print("\n")

print("Vector r: ", r)
print("Order of r: ", r.shape)
print("\n")

print("Multiplied matrix result:")
print(res)
print("Order of multiplied matrix: ", res.shape)

Original matrix A:
[[4 5 6]
 [7 8 9]]
Order of A:  (2, 3)


Vector r:  [1 2]
Order of r:  (2,)


Multiplied matrix result:
[[18 21 24]]
Order of multiplied matrix:  (1, 3)


In [127]:
# Multiplication with a column vector
import numpy as np

A = np.asmatrix(np.array([ # 2 x 3 matrix
    [1, 2, 3],
    [4, 5, 3]
]))
r = np.array([[4], [5], [6]])   # col vector 3 x 1
res = np.matmul(A, r) # matrix multiplication

print("Original matrix A:")
print(A)
print("Order of A: ", A.shape)
print("\n")

print("Vector r: ", r)
print("Order of r: ", r.shape)
print("\n")

print("Multiplied matrix result:")
print(res)
print("Order of multiplied matrix: ", res.shape)

Original matrix A:
[[1 2 3]
 [4 5 3]]
Order of A:  (2, 3)


Vector r:  [[4]
 [5]
 [6]]
Order of r:  (3, 1)


Multiplied matrix result:
[[32]
 [59]]
Order of multiplied matrix:  (2, 1)


### Matrix Multiplication

A matrix A can be multiplied by a matrix B. Some properties of matrix multiplications are:

1. (A . B) . C == A . (B . C)
2. A . (B + C) == (A . B) + (A . C)
3. A . B != B . A


In [140]:
# Multiplication of two matrices
import numpy as np

A = np.asmatrix(np.array([ # 2 x 3 matrix
    [1, 2, 3],
    [4, 5, 3]
]))
B = np.asmatrix(np.array([ # 3 x 2 matrix
    [4, 5],
    [5, 3],
    [6, 1],
])) 

res = np.matmul(A, B) # matrix multiplication

print("Matrix A:")
print(A)
print("Order of A: ", A.shape)
print("\n")

print("Matrix B:")
print(B)
print("Order of B:", B.shape)
print("\n")

print("Multiplied matrix result:")
print(res)
print("Order of multiplied matrix: ", res.shape)

Matrix A:
[[1 2 3]
 [4 5 3]]
Order of A:  (2, 3)


Matrix B:
[[4 5]
 [5 3]
 [6 1]]
Order of B: (3, 2)


Multiplied matrix result:
[[32 14]
 [59 38]]
Order of multiplied matrix:  (2, 2)


## Transposition of Matrices

A matrix of order m x n can be transposed to a matrix At of order n x m where At(i, j) = A(j, i)
Some properties regarding Transposition of matrices:

1. Att == A, where Att is tranposition of At. Double transposition of A
2. (A + B)t == At + Bt
3. (cA)t == c(At), where c is a scalar quantity
4. (A . B)t == Bt . At ---> product of transposed matrices
5. Tr(A) == Tr(At), where Tr(A) is the <strong>Trace</strong> of a matrix A, i.e., sum of all the diagonal elements.
6. inv(At) == (inv(A))t, where inv(A) is the inverse of matrix A.


The transposition of matrices brings us two more special types of matrices:

1. <strong>Symmetrical Matrix: </strong>When At == A, the matrix is said to be symmetrical.
2. <strong>Skew-Symmetrical Matrix: </strong>When At == -A, the matrix is said to be skew-symmetrical.


In [111]:
# Symmetrical Matrix
import numpy as np

# Create a symmetric matrix
A = np.asmatrix(np.array([
    [1, 2, 3],
    [2, 4, 5],
    [3, 5, 6]
]))
At = np.transpose(A)

# Print the original matrix
print("Original Matrix:")
print(mat)
print("\n")

# Print the transposed matrix
print("Transposed Matrix:")
print(At) # observe that At == A

Original Matrix:
[[1 2 3]
 [2 4 5]
 [3 5 6]]


Transposed Matrix:
[[1 2 3]
 [2 4 5]
 [3 5 6]]


In [114]:
# Skew Symmetrical Matrix
import numpy as np

# Create a symmetric matrix
A = np.asmatrix(np.array([
    [0, 2, -3],
    [-2, 0, 4],
    [3, -4, 0]
]))
At = np.transpose(A)

# Print the original matrix
print("Original Matrix:")
print(A)
print("\n")

# Print the transposed matrix
print("Transposed Matrix:")
print(At) # observe that At == -A

Original Matrix:
[[ 0  2 -3]
 [-2  0  4]
 [ 3 -4  0]]


Transposed Matrix:
[[ 0 -2  3]
 [ 2  0 -4]
 [-3  4  0]]


## Determinant

The determinant of a matrix is it's magnitude. Just like |v| for a vector v. Or, the total area scaled by a matrix during linear transformation.


In [145]:
# Calculating the determinant
import numpy as np


A = np.arange(1, 10).reshape(3, 3)
print("Matrix A:")
print(A)

detA = int(np.linalg.det(A)) # gives a number with e not an int so convert the determinant to an int
print("Determinant of A: ", detA)

Matrix A:
[[1 2 3]
 [4 5 6]
 [7 8 9]]
Determinant of A:  0


## Minors and CoFactors

Let a 3 x 3 Matrix A = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]. The minor of each element A(i, j) is the determinant of the matrix made by discarding ith row and jth column.</br>
Example - M(0, 0), minor of the first element would be det([[5, 6], [8, 9]]). And the cofactor C(i, j) = -1^(i + j) \* M(i, j)</br>
</br>
Minors can be calcuated for each element of the matrix and hence, they can be a matrix of Minors where M(i, j) = minor(A(i, j)).</br>
</br>
Cofactors can be calcuated for each element of the matrix and hence, they can be a matrix of Cofactors where C(i, j) = -1^(i + j) \* M(i, j).</br>


In [152]:
# Calculating minors of a matrix
from sympy import Matrix # numpy doesn't provide functions to calculate minors. We can create our own functions but sympy is more simple

# Create a function to calculate the minor matrix
def get_minor_matrix(A):
    rows, cols = A.shape
    minor_ls = [[A.minor(i, j) for j in range(cols)] for i in range(rows)] # calculate the minor of each element

    return Matrix(minor_ls)

# Create a symbolic matrix
A = Matrix([[1, 2, 3],
            [4, 5, 6],
            [7, 8, 9]])

# Calculate the minors matrix
minors_A = get_minor_matrix(A)

# Print the result
print("Matrix A:")
print(A)
print("\n")

print("Minors of Matrix A:")
print(minors_A)

Matrix A:
Matrix([[1, 2, 3], [4, 5, 6], [7, 8, 9]])


Minors of Matrix A:
Matrix([[-3, -6, -3], [-6, -12, -6], [-3, -6, -3]])


In [154]:
# Calculating cofactors of a matrix
from sympy import Matrix

# Create a symbolic matrix
A = Matrix([[1, 2, 3],
            [4, 5, 6],
            [7, 8, 9]])

# Calculate the cofactor matrix
cofactor_A = A.cofactor_matrix()

# Print the result
print("Matrix A:")
print(A)
print("\n")

print("Cofactors of Matrix A:")
print(cofactor_A)

Matrix A:
Matrix([[1, 2, 3], [4, 5, 6], [7, 8, 9]])


Cofactors of Matrix A:
Matrix([[-3, 6, -3], [6, -12, 6], [-3, 6, -3]])


## Adjoint or Adjugate of a matrix

Adjoint matrix of matrix A is Adj(A). Let a matrix K created by using A such that K(i, j) = Cofactor(A(i, j)). The transpose of this matrix K is called adjoint matrix.


In [155]:
# Calculating adjoint of a matrix
from sympy import Matrix

# Create a symbolic matrix
A = Matrix([[1, 2, 3],
            [4, 5, 6],
            [7, 8, 9]])

# Calculate the adjoint matrix
adjointA = A.adjugate()

# Print the result
print("Matrix A:")
print(A)
print("\n")

print("Adjoint of Matrix A:")
print(adjointA)

Matrix A:
Matrix([[1, 2, 3], [4, 5, 6], [7, 8, 9]])


Adjoint of Matrix A:
Matrix([[-3, 6, -3], [6, -12, 6], [-3, 6, -3]])


## Inverses

The inverse of a m x n matrix A is denoted by A^-1 such that A \* A^-1 = I, where I is the identity matrix.

- Not all matrices has inverses.
- Inverse of a matrix A only exists if det(A) != 0 (The matrix should not be singular).
- Inv(A) = Adj(A) / det(A).

Some properties of inverse of matrices:

1. A . inv(A) == inv(A) . A == I
2. I == inv(I)
3. (inv(A))t == inv(At)
4. inv(A . B) == inv(B) . inv(A)
5. inv(k . A) == 1/k . inv(A)


In [157]:
# Finding the non existent inverse of a matrix ---> the matrix is singular i.e, det(A) = 0
import numpy as np

A = np.arange(2, 11).reshape(3, 3)
print(A)

# A = np.array([[1, 3], [4, 5]])
detA = int(np.linalg.det(A))
print("Determinant:", detA)

print("Inverse: ")
print(np.linalg.inv(A))

[[ 2  3  4]
 [ 5  6  7]
 [ 8  9 10]]
Determinant: 0
Inverse: 


LinAlgError: Singular matrix

In [158]:
# Finding the inverse of a non-singular matrix, i.e, determinant is not 0
A = np.array([[2, 4], [1, 3]])
print(A)
print("\n")

detA = int(np.linalg.det(A))
print("Determinant:", detA)

print("Inverse: ")
print(np.linalg.inv(A))

[[2 4]
 [1 3]]


Determinant: 2
Inverse: 
[[ 1.5 -2. ]
 [-0.5  1. ]]


## Echelon Form (Row Echelon Form) and Rank of a Matrix

The Row Echelon Form of a matrix A is an upper right triangle matrix. It's a special form of a matrix that gives us how much information it has.


In [161]:
# Calculating the Row Echelon form of a matrix
from sympy import Matrix

A = Matrix([
    [2, 0, 2, 6],    # nothing but 2 * R3 (row 3) of the matrix
    [2, 3, 4, 7],
    [1, 0, 1, 3],
])
RRefA = A.rref() # calculates the row echelon form of the matrix

print("Matrix A:")
print(A)
print("\n")

print("Row Echelon Form of the matrix A:")
print(RRefA)
print("\n")

print("Rank of the matrix A:", A.rank())

Matrix A:
Matrix([[2, 0, 2, 6], [2, 3, 4, 7], [1, 0, 1, 3]])


Row Echelon Form of the matrix A:
(Matrix([
[1, 0,   1,   3],
[0, 1, 2/3, 1/3],
[0, 0,   0,   0]]), (0, 1))


Rank of the matrix A: 2


## Eigen Values and Eigen Vectors

For a matrix A, such that (A - pI) \* v = 0, if there exists at lease one value of p, and one vector v to satisfy the equation. The value p is called Eigen Value and the corresponding vector v is called Eigen Vector.

- For an Eigen Vector to exist for a matrix A det(A - pI) == 0.
- There can be more than one Eigen Values and correspoding Vectors for which the condition is satisfied.
- The maximum number of Eigen Value and Eigen Vectors is the number or rows of the square matrix.
- An identity matrix has only one eigen value and that is 1


In [165]:
# Calculating Eigen values and Eigen Vectors of a matrix
from sympy import Matrix
A = Matrix([[2, 1], [-3, 6]])

evecs = A.eigenvects() # --> prints [(scaleX, scaleY), Eigen Vectors] --> scaleX and scaleY is nothing but the eigen values along x and y

for scaleX, scaleY, vector in evecs:
    print("Eigen Values:", scaleX, scaleY)
    print("Eigen Vector:")
    print(vector)
    print("\n")

Eigen Values: 3 1
Eigen Vector:
[Matrix([
[1],
[1]])]


Eigen Values: 5 1
Eigen Vector:
[Matrix([
[1/3],
[  1]])]




# Linear Algebra

Now we go through some linear algebra concepts used in machine learning


# System of Linear Equations

There are 2 types of Equations


## Two types of Linear Equations

1. Singular - The determinant of this type of system is always 0.
2. Non-singular - The determinant of this type of system is always a real number other than 0.


### Singular

A set of equations which has infite number of solutions or no solution is called a singular system.

1. Infinite solutions - 2x + 3y = 1, 4x + 6y = 2
2. No Solution - 2x + 3y = 1, 2x + 3y = 0


In [167]:
# Representing a singular system and calcualating the determinant
import numpy as np

A = np.asmatrix(np.array([
    [2, 3], # --> 2x + 3y
    [4, 6], # --> 4x + 6y
]))
detA = int(np.linalg.det(A))

print("Matrix A:")
print(A)
print("Determinant of A: ", detA)

Matrix A:
[[2 3]
 [4 6]]
Determinant of A:  0


### Non-singular

A set of equations which can be solved and has a unique solution is called a non-singular system.</br>
Ex: 2x + 3y = 13, 4x + y = 11 where x = 2 and y = 3.</br>
The determinant of this system will always be a real number.</br>


In [168]:
# Representing a non-singular system and calcualating the determinant
import numpy as np

A = np.asmatrix(np.array([
    [2, 3], # --> 2x + 3y
    [4, 1], # --> 4x + y
]))
detA = int(np.linalg.det(A))

print("Matrix A:")
print(A)
print("Determinant of A: ", detA)

Matrix A:
[[2 3]
 [4 1]]
Determinant of A:  -10


## Matrix Representation

All the linear system of equations can be represented as matrices.</br>


2x + 3y = 13, 4x + y = 11 can be represented by:</br>
</br>

- A = [[2, 3], [4, 1]]
- X = [[x], [y]]
- Y = [[13], [11]]

Here the equation translates to AX = Y

1. Multiply both sides by inv(A) or inverse(A) -> inv(A) x A x X = inv(A) x Y
2. We get -> X = inv(A) x Y
3. We know how det(A) is used to calculate inv(A). Hence, for the system to be solvable the determinant != 0 and the inverse must exist.


Question: Calculate the Solution for</br>
2x + 3y = 13</br>
4x + y = 11 </br>
</br>
Answer:

1. Represent this in matrix form: A = [[2, 3], [4, 1]], X = [[x], [y]], Y = [[13], [11]]
2. The solution for the system is X = inv(A) x Y

Calulations below


In [169]:
import numpy as np

A = np.array([[2, 3], [4, 1]])
# X = [[x], [y]]
Y = np.array([[13], [11]])

X = np.matmul(np.linalg.inv(A), Y)
print(X)

[[2.]
 [3.]]


Question: Calculate the Solution for</br>
2x + 3y + z = 11</br>
4x + y + 2z = 12</br>
3x + 2y = 7 </br>
</br>
Answer:

1. Represent this in matrix form: A = [[2, 3, 1], [4, 1, 2], [3, 2, 0]], X = [[x], [y], [z]], Y = [[11], [12], [7]]
2. The solution for the system is X = inv(A) x Y

Calulations below


In [170]:
import numpy as np

A = np.array([[2, 3, 1], [4, 1, 2], [3, 2, 0]])
# X = [[x], [y], [z]]
Y = np.array([[11], [12], [7]])

X = np.matmul(np.linalg.inv(A), Y)
print(X)

[[1.]
 [2.]
 [3.]]


## Linear Dependence and Indpendence


For a system of equations say E1, E2, and E3

1. If there exists a value such that a \* E1 = E2, the we say E2 is dependent on E1 - they both are the same equations and hence, infinite solutions exist.
   </br>
   Example:</br>
   E1 -> 2x + 4y = 10</br>
   E2 -> 4x + 8y = 20</br>
   </br>
   Here E2 = 2 x E1. Hence the system is linearly dependent.

2. Or, E3 = a \* (E1 + E2) -> E3 is dependent on both E1 and E2.
   </br>
   Example:</br>
   E1 -> x + y = 5</br>
   E2 -> x + 2y = 8</br>
   E3 -> 4x + 6y = 26</br>
   </br>
   Here E3 = 2 x (E1 + E2). Hence the system is linearly dependent.</br>
   </br>
   <strong>Note: </strong> The determinant of a linearly dependent system of equations will always be 0


## Rank of a Matrix and Echelon Form

The Echelon Form (Row Echelon and Column Echelon) of a matrix tells us how much information it has.

- Rank of a matrix means how much information it has. For ex, if there are 3 equations (e1, e2, e3) but 2 (e2 and e3) are similar then only one of them (either e2 or e3) is actually useful to us, hence the rank of the matrix is 2.
- Applications include compressing images (SVD method) to store the same information but in a reduced way - blurred or low quality image.


### Echelon Form of a Matrix - Row Echelon Form

The Echelon form of a Matrix A is a special type of upper-right triangle matrix which has following properties:

1. A(0, 0) == 1 holds true.
2. A(1, 0) == 0 and A(1, 1) == 0 or 1 (usually 1)
   </br></br>
3. In general for an element at position (i, j) of a matrix:
   </br></br>
   1. A(i, j) == 0 for (i > j).
   2. A(i, j) == a where a is the first element of row i. a can be 0. a != 1 if it is row echelon form of the matrix. a == 1 if it is <strong>reduced</strong> row echelon form of the matrix.
   3. A(i, j) == b for (i < j) where b is any real number.
   4. If A(i, j) == 0 for any i the i must be equal to m-1 (last row).


#### Pivot

A pivot of a row is the first non-zero value in the row.
</br></br>
Apply sclar, row, or column (either row or column only) operations on all the elements of a matrix to get it's Echelon form.</br>
</br>
Given a matrix A:</br>
[[1, 2, 3, 1],</br>
 [2, 4, 6, 2],</br>
 [1, 2, 2, 2]]</br>
</br>
Apply following transformations in order to get Row Echelon Form:</br>

1. R1 -> R1 - 2R0 This will get 0 on A(1, 0), i > j (condition 1) but will result in 0, 0, 0, 0 on R1
2. R2 -> R2 - R0 This will get 0 on A(2, 0) and A(2, 1), i > j (condition 1)
3. R1 <-> R2 This will make R2 all 0s (condition 4)
4. R1 -> (-1)\*R1 This will make A(1, 2) as 1 (condition 2)
   </br>
   </br>

It will result in the following matrix:
</br></br>
[[1, 2, 3, 1],</br>
 [0, 0, 1, -1],</br>
 [0, 0, 0, 0]]</br>

</br>
This is the Reduced Row Echelon form of the Matrix A where all pivots are 1.


In [171]:
# finding the echelon form of a matrix - we use sympy module for this
from sympy import Matrix

A = Matrix([
    [1, 2, 3, 1],
    [2, 4, 6, 2],
    [1, 2, 2, 2]
])

rE = A.rref()
print(rE)

(Matrix([
[1, 2, 0,  4],
[0, 0, 1, -1],
[0, 0, 0,  0]]), (0, 2))


### Rank of the Matrix A

Rank means how many unique information does the system have.


#### Order of a Matrix A

1. If m == n then the order of the matrix ord(A) = m. Else, the order of the matrix ord(A) = m x n.
2. Order is represented as m x n and it is not a product or m and n.
3. For all the below examples we are considering only the row order i.e., the value of m.

#### Rank and Order

If det(A) != 0 the the rank(A) = ordr(A) - This means all the information about the equations are unique.</br>
</br>
If det(A) == 0 then:

1. rank(A) <= ordr(A) for all A whose det(A) == 0
2. rank(A) = number of non-zero rows in the row echelon form of the matrix


In [4]:
# Find the rank of the matrix
import numpy as np
A = np.array([
    [1, 2, 3, 1],
    [2, 4, 6, 2],
    [1, 2, 2, 2]
])

rank = np.linalg.matrix_rank(A)
print(rank)

2


## Matrices as Linear Transformations

### Some Key Terms

<strong>Basis: </strong>The square formed by the points (0, 0), (0, 1), (1, 0), and (1, 1) is called the basis.</br>
<strong>Transformation: </strong>When one plane system is translated into another via a transformation matrix it is called Transformation</br>
<strong>Transformation Matrix: </strong>A matrix translates one basis to another.</br>
<strong>Span: </strong>The span of a vector is a line passing through origin and the vector i.e., the line on which the vector lies.

- The span has two components, a line and the angle.
- The line equation can be calulated by y = mx + b where m is the slope and b is the y intercept. Put (0, 0) in the equation to get the value of b.
- The angle can be calculated by using tan-1(m).


### How transformation works.


Original Basis = (0, 0), (0, 1), (1, 0), and (1, 1)
Transformation Matrix = [[3, 1], [1, 2]]

We multiply the transformation matrix with each point in the Original Basis to get a new basis.

- (0, 0) -> [[3, 1], [1, 2]] x [[0], [0]] = [[0], [0]] = (0, 0) ---> origin always gets translated to origin.
- (0, 1) -> [[3, 1], [1, 2]] x [[0], [1]] = [[1], [2]] = (1, 2)
- (1, 0) -> [[3, 1], [1, 2]] x [[1], [0]] = [[3], [1]] = (3, 1)
- (1, 1) -> [[3, 1], [1, 2]] x [[1], [1]] = [[4], [3]] = (4, 3)

These are the co-ordinates of the new basis.</br>
Similarly any vector (a, b) can be translated to the new basis by:</br>

- (a, b) -> [[3, 1], [1, 2]] x [[a], [b]] = [[p], [q]] = (p, q)


### Matrix to Vector Multiplication as Tranformations


We only need two vectors u = (0, 1) and v = (1, 0) and a Transformation matrix A to get the new Matrix.

- We multiply A and u to get the tu (transformed u: the (0, 1) in the new plane)
- We multiply A and v to get the tv (transformed v: the (1, 0) in the new plane)
- Then we add tu and tv to get the (1, 1) in the new plane.
- The (0, 0) remains at origin.

By using these 4 points we get the new basis.
<br/></br>
Let's say we have a matrix A and a matrix B. The matrix A transforms the original space to sA (some abstract notation to denote transformation) then we use the matrix B to further transform the sA to sB. Is there any matrix C such that it directly transforms the space s to sB? Yes, the C is A x B.</br>
</br>
i.e., C = np.matmul(A, B)


### Intuition behind Eigen Values and Eigen Vectors


After transforming a vector to a new basis using the transformation matrix (i.e., multiplying the matrix with the vector), the original vector will no longer follow its span. It will get knocked off.</br>
However, for a particular transformation matrix, some vectors do follow their own span and just get scaled down or up in the new basis.</br>
</br>
Example 1: For a transformation matrix A = [[3, 1], [0, 2]] the vector u = (1, 0) translates to (3, 0) i.e., A x u = (3, 0).</br>
It is obvious that the vector span remains the same (i.e., along the x-axis) but it is scaled by a factor of 3.
</br></br>
Example 2: For the same transformation matrix A = [[3, 1], [0, 2]] the vector v = (-1, 1) translates to (-2, 2) i.e., A x v = (-2, 2).</br>
Clearly, the span remains the same but the vector is scaled by a factor of 2.

#### Eigen Vectors

The special vectors for a particular transformation matrix for which the span remains the same are called Eigen Vectors.

#### Eigen Values

The factor by which the Eigen Vectors are scaled (stretched or squished) is called Eigen Values.</br>

#### Eigen Basis

The basis after the transformation. It is the matrix formed by stacking the eigen vectors together.</br>
</br>
<strong>Note: </strong>For an nD space, the span will be line only. The span represents a line that extends from origin to infinity in an nD space on which the vector lies.
[Know More](https://www.youtube.com/watch?v=PFDu9oVAE-g&t=50s)
</br></br>
Below code demonstrates this with the same transformation matrix A.


In [174]:
# Printing Eigen Values and Vectors 
from sympy import Matrix
A = Matrix([
    [2, 0, 0],
    [1, 2, 1],
    [-1, 0, 1]
])

evecs = A.eigenvects() # --> prints [(scaleX, scaleY), Eigen Vectors] --> scaleX and scaleY is nothing but the eigen values along x and y

for scaleX, scaleY, vector in evecs:
    print("Eigen Values:", scaleX, scaleY)
    print("Eigen Vector:")
    print(vector)
    print("\n")

Eigen Values: 1 1
Eigen Vector:
[Matrix([
[ 0],
[-1],
[ 1]])]


Eigen Values: 2 2
Eigen Vector:
[Matrix([
[0],
[1],
[0]]), Matrix([
[-1],
[ 0],
[ 1]])]




### Intuition behind Determinants


[Watch this for Clarity](https://www.youtube.com/watch?v=Ip3X9LOh2dk)

The determinant of a matrix is the final number by which the area of the basis of a space is scaled after applying the transformation matrix or (simple matrix) to it.</br>
</br>
In simpler terms, determinant is the factor by which the area covered by rectangle (0, 0), (0, 1), (1, 0), (1, 1) [basis] is scaled by the matrix.</br>
</br>
If a matrix A scales the basis by 5 (i.e., det(A) is 5), and det(B) is 3 then it is clear that the matrix A will scale area by 5 and then matrix B will further scale the area by 3 so the total scaling would be det(A) x det(B).</br>
</br>
When we multiply (cross-product) two matrices we essentially calculating a matrix C to go from initial space P to final space R directly without doing P to Q by applying A and then Q to R by applying B. Therefore, the matrix C will have the same scaling as scaling of A then B. Hence, the det(AxB) = det(A) x det(B).

- The scaling of a singular matrix A i.e., det(A) == 0.
- Therefore det(AB) = det(A) x det(B) == 0. Hence, total scaling is also 0.
  </br></br>

<strong>Note: </strong>The scaling of an inverse matrix is 1/scaling of the original matrix. i.e. det(inv(A)) = 1/det(A)


### Intuition behind Singularity and Rank (and Echelon Form)


If a matrix (or transormation matrix) is singular (i.e., the determinant is 0), the transformation matrix actually reduces the number of dimensions in the space.</br>
Example: If a determinant of the matrix A is 0, then it can reduce a 3D space to a 2D or a 1D space.</br>
</br>
The number of dimensions that we will get after reduction is determined by the rank of the matrix (which can be easily calculated by Row Echelon form of the matrix).


### Another way to find inverse of a Matrix


Let A = [[1, 3], [4, 5]]</br>
inv(A) = [[a, b], [c, d]]</br>
</br>
Now we know that inv(A)_A = I (i.e., [[1, 0], [0, 1]]). inv(A) _ A gives us following equations:

1. a + 3c = 1
2. b + 3d = 0
3. 4a + 5c = 0
4. 4b + 5d = 1
   </br></br>

We get a = 1.5, b = -2, c = -0.5, and d = 1


### PCA - Principal Component Analysis

Dimensionality Reduction


In a dataset we often get some rows or columns that are linearly dependent. We can remove such rows or columns from the data set to make our training more efficient and fast.</br></br>
The second way to reduce our data such that it still contains useful information is by doing Principal Component Analysis. Suppose we have a dataset of two points (x, y) that lies approximately in the line L => y = mx + b. We can approximate those points to exactly fit the line L, so that we can calculate y from x thereby reducing one row/column of the data.</br>
</br>
This method is called dimensionality reduction and it is done via PCA algorithm and it works on the principles of Eigen Values and Eigen Vectors. The dimensionality reduction is done to make the manipulation and training of model easier with retaining as much information as we can.


# Simple Machine Learning Classifier with Matrices

Spam Mail Classification


## Dataset

We create the following table to determine (by ourselves, this is training data) that a mail is spam or not based on the count of the two words i.e., "Lottery" and "Win".</br>
| Lottery | Win | Total Score? | Spam |
| ------- | --- | ------------ | ---- |
| 1 | 1 | 2 | Yes |
| 2 | 1 | 3 | Yes |
| 0 | 0 | 0 | No |
| 0 | 2 | 2 | Yes |
| 0 | 1 | 1 | No |
| 1 | 0 | 1 | No |
| 2 | 2 | 4 | Yes |
| 2 | 0 | 2 | Yes |
| 1 | 2 | 3 | Yes |

Then we determine a threshold that satisfy all these conditions. For ex: a <strong>threshold of 1.5</strong> will satisfy all conditions. If the score is above 1.5, the email maked as spam. By looking at the data available to us above, we can also use 1.9 as the threshold as it will also cover the same conditions. Choosing threshold is an attentive tasks and most libraries have internal formulas to choose it automatically.</br>


## The Model

In the training process the model comes up with the following matrix (we are not actually training the model, just for demonstration purposes)

- [[1], [1]] - result matrix (we are giving equal weights to both the words)
- The model calculates a threshold of 1.5 (or 1.9, 1.7, etc)
- The model accepts two inputs "Win" count and "Lottery" count in the form of a vector [L, W] from the processed natural languages.
- Then the model calculates the score of the input [L, W] \* [[1], [1]], compares it with the threshold, and outputs with (1, 0) (spam, no spam).

Example:</br>
Row 1 -> [1, 1] x [[1], [1]] = 1x1 + 1x1 = 2 > Threshold (i.e., 1.5) ---> Spam</br>
Row 6 -> [1, 0] x [[1], [1]] = 1x1 + 1x0 = 1 < Threshold (i.e., 1.5) ---> Not Spam

#### In reality, we ourselves need to train the models and they are much more complicated.


## The Bias

Let's say we apply a bias of -1.5 to the score of the model. Then we only need to check if the resultant score (actual score from model - 1.5) is positive or negative.

- Positive? Spam
- Negative? Not spam
